### DISCUSSION TITLE
Is there a recommended way to create and use EVENT_TABLE

### DISCUSSION THREAD

**Discussion by eliavra** [Snowflake-Labs/terraform-provider-snowflake#2166] [Q&A] [answered]
Hey there! I'm hoping you can help me out with something. I'm looking to formalize the creation and usage of the EVENT_TABLE through Terraform. Do you think it's possible to do so using the Snowflake provider? Thank you!

```
CREATE EVENT TABLE
ALTER ACCOUNT SET EVENT_TABLE
```

**Comment by sfc-gh-asawicki** [comment, marked as answer]
Hey, @eliavra. Thank you for reaching out to us.

We do not yet support event tables, but ultimately, we will support them. @sfc-gh-swinkler is currently working on event table support in our SDK, but this is just the first thing that must be done to support the event table in the provider. Later, we will still have to add it as the new resource/datasource; this probably won't happen before the first quarter of the following year.

**Reply by eliavr** [reply]
Thank you @sfc-gh-asawicki  for your elaborate answer. 
Can you suggest a workaround, like running a SQL query through Terraform?

**Reply by sfc-gh-asawicki** [reply]
There are other providers like https://registry.terraform.io/providers/paultyng/sql/latest/docs or https://registry.terraform.io/providers/aidanmelen/snowsql/latest/docs which allow you to write custom SQL statements.

At some point, we will also introduce a generic SQL resource to run pure SnowSQL statements/queries. We are not there yet, though.

Another option is to run scripts outside Terraform: I imagine that terraform plan/apply is run as one of your CI/CD pipeline steps so that you can add custom scripts in the following step. This is not the endgame solution, but it can serve as a workaround now.

**Reply by eliavra** [reply]
Thank you for your help! 
I will use one of the suggested solutions for our project.

**Comment by MyJBMe** [comment]
Hey @sfc-gh-asawicki, 

is there an update on this topic? Stating "probably won't happen before the first quarter of the following year" is now over two years old and for me quite an important topic right now :)

**Reply by sfc-gh-jcieslak** [reply]
Hey @MyJBMe 
Currently, there's no event table resource (will be in https://registry.terraform.io/providers/snowflakedb/snowflake/latest/docs, most likely under the `snowflake_event_table` name). Currently, you can set an event table on an account through Terraform using either `account_parameter` or `current_account` resources. The even table can be created through Terraform using the [execute](https://registry.terraform.io/providers/snowflakedb/snowflake/latest/docs/resources/execute) resource, which is a more manual approach and limited in its ability to detect and react to external changes. If you would like to emphasize the importance of providing support for the event table, please collaborate with your point of contact in Snowflake to submit a product gap for our project.

**Reply by sfc-gh-jcieslak** [reply]
Additionally, although we have already set our plans for this quarter, we will accept and assist if you would like to contribute to adding support for the event table resource. We have already completed part of the implementation some time ago, and the resource can be implemented with limited capabilities.

**Reply by MyJBMe** [reply]
Would be happy to help! Anything to read (besides contribution section) before getting started? Feels like anything about SDK is important, but as some links (generator readme) are broken there, could need a little head start :)

**Reply by sfc-gh-jcieslak** [reply]
I'll try to respond tomorrow. We already have some of the SDK implemented for even tables, but I have to check how up to date it is, etc. I'll also check if we had a similar answer on what has to be done and give some example PRs.

**Reply by sfc-gh-jcieslak** [reply]
Hi again,
It appears that the event table is largely covered in the SDK, so I think we can skip this part and proceed directly to resource implementation. Here's a good example of implementing a resource: https://github.com/snowflakedb/terraform-provider-snowflake/pull/3739. You have to create a new event table file within the resources package, implement CRUD operations in it (for now, let's focus on very simple ones), generate assertions using https://github.com/snowflakedb/terraform-provider-snowflake/blob/main/pkg/acceptance/bettertestspoc/README.md, and write acceptance tests following pattern set in this PR: https://github.com/snowflakedb/terraform-provider-snowflake/pull/4113 (will be described within the repo soon). Most of the things you can achieve by copying the linked PR, but in case you do not know how to approach certain things, feel free to reach out to us either through a dedicated issue, here, or directly in the PR.

**Reply by MyJBMe** [reply]
Hey there, I checked your resources and implemented everything, needs some testing now. Before I finish with tests, I wanted to make sure how you set up all SDK files. Are these completely generated files? I felt like RLS would be nice as optional feature (took VIEW resource as base for possible parameters, as already stable), but event tables existing SDK implementation did not offer any helpful bases. I thereby added those manually (copied from other SDK implementations and corrected types accordingly). Nonetheless it bothered me seeing this gentle DO NOT TOUCH comment at the top of each SDK file. Could you give me some insights into a correct process here? Struggled a little when generating SDK files...

**Reply by sfc-gh-jcieslak** [reply]
The sdk objects are generated from the `pkg/sdk/generator/defs/*_def.go` files (`pkg/sdk/generator/defs/event_tables_def.go` in this case). For more info on generating sdk files you can read `pkg/sdk/generator/README.md`. To add new functionality you have to map SQL syntax to the definition in def, then regenerate the SDK object. The information at the top of the file is pretty standard for generated files in Go. The only difference in our case is that some of the files still require manual adjustments (we usually comment them with `manually adjusted` comment, but there may be missing places as it wasn't the practice from the start). Ultimately, we aim to distinguish between auto-generated content and the sections that require manual adjustments, but that's where we stand currently. I suggest to add a new functionality to the SDK (and then to the resource) in a follow-up PR. Adding a new resource should already contain a lot of lines to review.

**Reply by sfc-gh-jcieslak** [reply]
Hey @MyJBMe 👋 
How is the implementation going? Is there anything we can help you with? Remember that you don't have to aim for a fully complete PR; we can help you with the details.
